benches/classic_crypto.rs

```diff
@@ -0,0 +1,133 @@
+    /// 4. Choose `pk = 65537 = 0x10000000000000001` statically as public key (common as `enc` is very
+    /// efficient with this key, has sufficient size, and guarantees to have an inverse).
+    /// 5. Calculate `sk = pk^(-1) mod phi(N)`.
+    /// 6. Output `(N, pk, sk)`.
+    pub fn gen(security_lvl: u32) -> (Modulus, Z, Z) {
+        let lower_bound = Z::from(2).pow(security_lvl / 2).unwrap();
+        let upper_bound = Z::from(2).pow(security_lvl / 2 + 1).unwrap();
+
+        let p = sample_prime_naive(&lower_bound, &upper_bound);
+        let q = sample_prime_naive(&lower_bound, &upper_bound);
+        assert_ne!(p, q);
+
+        let modulus = Modulus::try_from(&(&p * &q)).unwrap();
+        let phi_mod = (&p - Z::ONE) * (&q - Z::ONE);
+
+        // standard prime value chosen as public key
+        // to remove necessity of choosing it at random
+        let pk = Z::from(65537);
+        let sk = Z::from(
+            Zq::try_from_z_z(&pk, &phi_mod)
+                .unwrap()
+                .inverse()
+                .expect("There must an inverse of this element as pk is prime."),
+        );
+
+        (modulus, pk, sk)
+    }
+
+    /// Returns a pre-computed RSA key pair that was computed with `gen(1024)`.
+    pub fn static_gen() -> (Modulus, Z, Z) {
+        let modulus = Modulus::from_str(RSA_N).unwrap();
+        let pk = Z::from(65537);
+        let sk = Z::from_str(RSA_PK).unwrap();
+        (modulus, pk, sk)
+    }
+
+    /// Encrypts a `msg` assumed to be smaller than `modulus` via RSA's `pk`.
+    ///
+    /// `RSA.Enc(N, pk, msg) = msg^pk mod N`
+    pub fn enc(modulus: &Modulus, pk: &Z, msg: &Z) -> Zq {
+        let msg = Zq::from_z_modulus(msg, modulus);
+        msg.pow(pk).unwrap()
+    }
+
+    /// Decrypts a `cipher` via RSA's `sk`.
+    ///
+    /// `RSA.Dec(N, sk, cipher) = cipher^sk mod N`
+    pub fn dec(sk: &Z, cipher: &Zq) -> Z {
+        Z::from(cipher.pow(sk).unwrap())
+    }
+
+    /// Run textbook RSA encryption with 1024 bit security.
+    /// 1. get (N, pk, sk) from a previously generated key pair with `gen(1024)`
+    /// 2. run cycle of `dec(sk, enc(pk, msg)) == msg`,
+    /// where `msg` is sampled uniformly at random in `[0, u64::MAX)`
+    pub fn rsa_run_enc_dec() {
+        let (modulus, pk, sk) = static_gen();
+        let msg = Z::sample_uniform(&0, &u64::MAX).unwrap();
+
+        let cipher = enc(&modulus, &pk, &msg);
+        let cmp = dec(&sk, &cipher);
+        assert_eq!(msg, cmp);
+    }
+
+    /// Run `RSA.gen` with security parameter `1024`, i.e. 1024-bit security.
+    pub fn rsa_run_gen() {
+        let _ = gen(1024);
+    }
+}
+
+/// benchmark [`rsa enc+dec`]
+pub fn bench_rsa_enc_dec(c: &mut Criterion) {
+    c.bench_function("RSA enc+dec", |b| {
+        b.iter(|| rsa_textbook::rsa_run_enc_dec())
+    });
+}
+
+/// benchmark [`rsa gen`]
+pub fn bench_rsa_gen(c: &mut Criterion) {
+    c.bench_function("RSA gen", |b| b.iter(|| rsa_textbook::rsa_run_gen()));
```

Comment:
Why did you not just call `rsa_textbook::gen(1024)` here?
```suggestion
    c.bench_function("RSA gen", |b| b.iter(|| rsa_textbook::gen(1024)));
```